src/dstack/_internal/server/background/tasks/process_runs.py

```diff
@@ -0,0 +1,283 @@
+import asyncio
+import datetime
+import itertools
+import uuid
+from typing import Iterable, List, Optional, Set, Tuple
+
+import sqlalchemy as sa
+from sqlalchemy.ext.asyncio import AsyncSession
+from sqlalchemy.orm import joinedload
+
+from dstack._internal.core.models.instances import InstanceOffer
+from dstack._internal.core.models.profiles import ProfileRetryPolicy
+from dstack._internal.core.models.runs import (
+    JobStatus,
+    JobTerminationReason,
+    RunSpec,
+    RunStatus,
+    RunTerminationReason,
+)
+from dstack._internal.server.db import get_session_ctx
+from dstack._internal.server.models import InstanceModel, JobModel, RunModel
+from dstack._internal.server.services.jobs import (
+    RUNNING_PROCESSING_JOBS_IDS,
+    RUNNING_PROCESSING_JOBS_LOCK,
+    SUBMITTED_PROCESSING_JOBS_IDS,
+    SUBMITTED_PROCESSING_JOBS_LOCK,
+    TERMINATING_PROCESSING_JOBS_IDS,
+    TERMINATING_PROCESSING_JOBS_LOCK,
+)
+from dstack._internal.server.services.runs import (
+    PROCESSING_RUNS_IDS,
+    PROCESSING_RUNS_LOCK,
+    create_job_model_for_new_submission,
+    fmt,
+    process_terminating_run,
+    run_model_to_run,
+)
+from dstack._internal.server.utils.common import wait_unlock
+from dstack._internal.utils.common import get_current_datetime
+from dstack._internal.utils.logging import get_logger
+
+logger = get_logger(__name__)
+PROCESSING_INTERVAL = datetime.timedelta(seconds=2)
+JOB_ERROR_CODES_TO_RETRY = {
+    JobTerminationReason.INTERRUPTED_BY_NO_CAPACITY,
+    JobTerminationReason.FAILED_TO_START_DUE_TO_NO_CAPACITY,
+}
+
+
+async def process_runs():
+    async with get_session_ctx() as session:
+        async with PROCESSING_RUNS_LOCK:
+            res = await session.execute(
+                sa.select(RunModel).where(
+                    RunModel.status.not_in(RunStatus.finished_statuses()),
+                    RunModel.last_processed_at < get_current_datetime() - PROCESSING_INTERVAL,
+                    RunModel.id.not_in(PROCESSING_RUNS_IDS),
+                )
+            )
+            runs: List[RunModel] = res.scalars().all()
```

Comment:
No need for List[RunModel]. Moreover, it's a wrong type.

Reply:
Why is it the wrong type?

I will remove it. In the first place, I've added it to help PyCharm with type hinting. Does VS Code accurately compute sqlalchemy return types?

Reply:
The defined return type is sequence: `def all(self) -> Sequence[_R]:`. Casting it to List raises Pylance warning.